pandas/core/case_when.py

```diff
@@ -0,0 +1,217 @@
+                f"however, it has a length {len(entry)}."
+            )
+    conditions, replacements = zip(*args)
+    common_dtype = [infer_dtype_from(replacement)[0] for replacement in replacements]
+
+    if default is not lib.no_default:
+        arg_dtype, _ = infer_dtype_from(default)
+        common_dtype.append(arg_dtype)
+    else:
+        default = None
+    if len(set(common_dtype)) > 1:
+        common_dtype = find_common_type(common_dtype)
+        replacements = [
+            construct_1d_arraylike_from_scalar(
+                value=replacement, length=len(condition), dtype=common_dtype
+            )
+            if is_scalar(replacement)
+            else replacement.astype(common_dtype)
+            if isinstance(replacement, ABCNDFrame)
+            else pd_array(replacement, dtype=common_dtype)
+            for condition, replacement in zip(conditions, replacements)
+        ]
+        if (default is not None) and isinstance(default, ABCNDFrame):
+            default = default.astype(common_dtype)
+    else:
+        common_dtype = common_dtype[0]
+    if not isinstance(default, ABCNDFrame):
+        cond_indices = [cond for cond in conditions if isinstance(cond, ABCNDFrame)]
+        replacement_indices = [
+            replacement
+            for replacement in replacements
+            if isinstance(replacement, ABCNDFrame)
+        ]
+        cond_length = None
+        if replacement_indices:
+            for left, right in zip_longest(
+                replacement_indices, replacement_indices[1:]
+            ):
+                if right is None:
+                    continue
+                if assert_index_equal(left.index, right.index, check_order=False):
+                    raise ValueError(
+                        "All replacement objects must have the same index."
+                    )
+        if cond_indices:
+            for left, right in zip_longest(cond_indices, cond_indices[1:]):
+                if right is None:
+                    continue
+                if assert_index_equal(left.index, right.index, check_order=False):
+                    raise ValueError("All condition objects must have the same index.")
+        if replacement_indices and cond_indices:
+            if assert_index_equal(
+                replacement_indices[0].index, cond_indices[0].index, check_order=False
+            ):
+                raise ValueError(
+                    "All replacement objects and condition objects "
+                    "should have the same index."
+                )
+        elif not cond_indices:
+            conditions = [
+                np.asanyarray(cond) if not hasattr(cond, "shape") else cond
+                for cond in conditions
+            ]
+            cond_length = {len(cond) for cond in conditions}
+            if len(cond_length) > 1:
+                raise ValueError("The boolean conditions should have the same length.")
```

Comment:
This is not true with alignment. It should only be enforced for non-indexed (I think non-Series?) objects.

Reply:
yes, this section applies to non-Series objects

Reply:
Ah, indeed, thanks.